proposals/NNNN-opt-in-reflection-metadata.md

```diff
@@ -0,0 +1,181 @@
+
+1.  Reflection Disabled (`-disable-reflection-metadata` and `-reflection-metadata-for-debugger-only`)
+- For Swift pre-6 emit Reflection metadata only if full debugging information is enabled.
+- If there is a type in a module conforming to `Reflectable`, the compiler will emit an error.
+- A no-op in Swift 6 and later (Opt-in mode is enabled by default).
+
+2.  Opt-In Reflection (`-enable-upcoming-feature OptInReflection`)
+- If debugging is disabled, emit only for types that conform to `Reflectable`.
+- Will emit reflection in full for all types if debugging is enabled.
+- For Swift pre-6 will require an explicit flag, for Swift 6 will be enabled by default.
+
+3.  Fully enabled (`-enable-full-reflection-metadata`)
+- Always emit reflection metadata for all types regardless of debugging information level.
+- Conformance to Reflectable will be synthesized for all types to allow usage of reflection-consuming APIs.
+- Current default level for Swift pre-6.
+
+Introducing a new flag to control the feature will allow us to safely roll it out and avoid breakages of the existing code. For those modules that get compiled with fully enabled metadata, nothing will change (all symbols will stay present). For modules that have the metadata disabled, but are consumers of reflectable API, the compiler will emit the error enforcing the guarantee.
+
+### Casts implementation
+Casting might be a good way to improve the feature's ergonomics because currently there is no way to check if reflection is available at runtime. (`Mirror.children.count` doesn't really help because it doesn't distinguish between the absence of reflection metadata and the absence of fields on a type)
+
+To implement this feature, we propose to introduce a new runtime function `swift_reflectableCast`, and emit a call to it instead of `swift_dynamicCast`during IRGen if Reflectable is a target type.
+
+Because of the fact that the compiler emits a call to that function at compile-time, all casts must be statically visible. All other cases like implicit conversion to `Reflectable` must be banned. This could be done at CSSimplify, when a new conversion constraint is introduced between a type variable and `Reflectable` type, the compiler will emit an error.
+
+```swift
+func cast<T, U>(_ x: U) -> T {
+    return x as! T
+}
+let a = cast(1) as Reflectable // expression can't be implicitly converted to Reflectable; use 'as? Reflectable' or 'as! Reflectable' instead
+let b: Reflectable = cast(1) // expression can't be implicitly converted to Reflectable; use 'as? Reflectable' or 'as! Reflectable' instead
+```
+Some diagnostics and optimizations will also have to be disabled even if conformance is statically visible to the compiler because all casts will have to go through the runtime call.
+
+**Availability checks**
+Since reflectable casting will require a new runtime function, it should be gated by availability checks. If a deployment target is lower than supported, an error will be emitted.
```

Comment:
We do have some support for backwards-deploying new runtime functions like this by adding them to the compatibility libraries.

Reply:
Is there going to be a compatibility library for Swift 6? (I am aware of them, but didn't want to introduce one specifically for this case)

Reply:
I think it's reasonable to assume there will be.

Reply:
Or at least, I think it makes sense to point out here that back-deployment would be possible if the new runtime function can be embedded in a compatibility library.